Src/IronPython/Compiler/Ast/ScopeStatement.cs

```diff
@@ -318,43 +256,47 @@ public abstract class ScopeStatement : Statement {
             throw new NotSupportedException();
         }
 
-        internal bool TryGetVariable(string name, out PythonVariable variable) {
-            if (Variables != null) {
+        internal bool TryGetVariable(string name, [NotNullWhen(true)] out PythonVariable? variable) {
+            if (Variables is not null) {
                 return Variables.TryGetValue(name, out variable);
             } else {
                 variable = null;
                 return false;
             }
         }
 
-        internal virtual bool TryBindOuter(ScopeStatement from, PythonReference reference, out PythonVariable variable) {
-            // Hide scope contents by default (only functions expose their locals)
+        internal virtual bool TryBindOuter(ScopeStatement from, PythonReference reference, [NotNullWhen(true)] out PythonVariable? variable) {
+            // Hide scope contents by default
             variable = null;
             return false;
         }
 
-        internal abstract PythonVariable BindReference(PythonNameBinder binder, PythonReference reference);
+        internal abstract PythonVariable? BindReference(PythonNameBinder binder, PythonReference reference);
 
         internal virtual void Bind(PythonNameBinder binder) {
-            if (_references != null) {
+            if (_references is not null) {
                 foreach (var reference in _references.Values) {
                     reference.PythonVariable = BindReference(binder, reference);
-                    if ((reference.PythonVariable is null || reference.PythonVariable.Kind is VariableKind.Global) && TryGetNonlocalStatement(reference.Name, out NonlocalStatement node)) {
+                    if ((reference.PythonVariable is null || reference.PythonVariable.Kind is VariableKind.Global) && TryGetNonlocalStatement(reference.Name, out NonlocalStatement? node)) {
                         binder.ReportSyntaxError($"no binding for nonlocal '{reference.Name}' found", node);
                     }
                 }
             }
         }
 
         internal virtual void FinishBind(PythonNameBinder binder) {
-            List<ClosureInfo> closureVariables = null;
+            List<ClosureInfo>? closureVariables = null;
+
+            if (FreeVariables is not null && FreeVariables.Count > 0) {
```

Comment:
Should we change this to `IsClosure`?

Reply:
Ah, I forgot that `MemberNotNullWhenAttribute` exists. Never used it before.